README.md
# binio

A Clojure library for reading and writing binary files.

## Usage

[![Clojars Project](https://img.shields.io/clojars/v/org.clojars.nighcoder/binio.svg)](https://clojars.org/org.clojars.nighcoder/binio)

Add the dependency [nighcoder/binio "0.1.1"] to your project.

Require the binio.core namespace.

`(require '[binio.core :as bio])`

The core namespace contains two functions to write to and read from files.

Spit-bytes writes a byte-array to file.

`(bio/spit-bytes "testfile" (byte-array [104 101 108 108 111 32 119 111 114 108 100 10]))`

Slurp-bytes reads a file and copies it's contents to a byte-array.

`(bio/slurp-bytes "testfile")`

## License

Copyright © 2020 Daniel Ciumberică

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED **AS IS**, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

## [0.1.1] - 2021-08-23
- Added decoding functions for various data types:
  - unsigned byte
  - signed short (16 bits)
  - unsigned short (16 bits)
  - signed long (32 bits)
  - unsigned long (32 bits)
  - signed long rational (64 bits)
  - unsigned long rational (64 bits)
  - single precision float (32 bits)

## [0.1.0] - 2020-01-27
### Released
- Released library with two functions: spit-bytes and slurp-bytes.


test/binio/core_test.clj
(ns binio.core-test
  (:require clojure.test
            [clojure.test.check :as tc]
            [clojure.test.check.generators :as gen]
            [clojure.test.check.properties :as prop]
            [clojure.test.check.clojure-test :refer [defspec]]
            [binio.core :refer :all]))


(defn ubyte?
  [n]
  (and (integer? n)
       (<= 0 n 255)))

(defn abs
  [x]
  (if (neg? x)
    (- x)
    x))

(defspec bits-roundtrip-test
  (do (println "Bits roundtrip test: Any number converted to bits and back remains unchanged")
    (prop/for-all [n (gen/choose 0 255)]
      (= (decode-bits (encode-bits n)) n))))

(defspec ubyte-test
  (do (println "Ubyte test: The decoded byte is an integer between 0 and 255 (inclusive)")
    (prop/for-all [b gen/byte]
      (ubyte? (decode-ubyte b)))))

(defspec ushort-sign-test 500
  (do (println "Ushort test: The decoded number is an integer between 0 and 65.535 (inclusive)")
    (prop/for-all [b (gen/vector gen/byte 2)]
      (and (integer? (decode-ushort b))
           (<= 0 (decode-ushort b) 65535)))))

(defspec ushort-sshort-conversion-test 500
  (do (println "Ushort Sshort conversion test: The decoded sshort converted to a ushort is the same as the decoded ushort")
    (prop/for-all [[b0 b1 :as b] (gen/vector gen/byte 2)]
      (if (neg? b0)
        (= (+ 65536 (decode-sshort b)) (decode-ushort b))
        (= (decode-ushort b) (decode-sshort b))))))

(defspec ulong-test 500
  (do (println "Ulong test: The decoded number is an integer between 0 and 4.294.967.295")
    (prop/for-all [b (gen/vector gen/byte 4)]
      (and (integer? (decode-ulong b))
           (<= 0 (decode-ulong b) 4294967295)))))

(defspec ulong-sslong-conversion-test 500
  (do (println "Ulong Slong conversion test: The decoded slong converted to a ulong is the same as the decoded ulong")
    (prop/for-all [[b0 & _ :as b] (gen/vector gen/byte 4)]
      (if (neg? b0)
        (= (+ 4294967296 (decode-slong b)) (decode-ulong b))
        (= (decode-ulong b) (decode-slong b))))))

(defspec ulong-rational-pos-test 500
  (do (println "Ulong rational positive test: The decoded number is a positive rational")
    (prop/for-all [b0 (gen/vector gen/byte 4)
                   b1 (gen/such-that #(not= 0 (reduce + %)) (gen/vector gen/byte 4))]
      (let [res (decode-ulongrational (concat b0 b1))]
        (and (pos? res) (rational? res))))))

(defspec ulong-rational-unit-test 500
  (do (println "Ulong rational unit test: The decoded number is > 1 when numerator > denominator")
    (prop/for-all [b0 (gen/vector gen/byte 4)
                   b1 (gen/such-that #(not= 0 (reduce + %)) (gen/vector gen/byte 4))]
      (let [res (decode-ulongrational (concat b0 b1))]
        (if (> (decode-ulong b0) (decode-ulong b1))
          (> res 1)
          (<= res 1))))))

(defspec slong-rational-test 500
  (do (println "Slong rational test: The decoded number is rational")
    (prop/for-all [b0 (gen/vector gen/byte 4)
                   b1 (gen/such-that #(not= 0 (reduce + %)) (gen/vector gen/byte 4))]
      (let [res (decode-ulongrational (concat b0 b1))]
        (rational? res)))))

(defspec slong-rational-sign-test 500
  (do (println "Slong rational sign test: The decoded number is positive if both the numerator and denominator are either positive or negative")
    (prop/for-all [b0 (gen/vector gen/byte 4)
                   b1 (gen/such-that #(not= 0 (reduce + %)) (gen/vector gen/byte 4))]
      (let [res (decode-slongrational (concat b0 b1))]
        (if (or (and (pos? (decode-slong b0)) (pos? (decode-slong b1)))
                (and (neg? (decode-slong b0)) (neg? (decode-slong b1))))
          (pos? res)
          (neg? res))))))

(defspec slong-rational-unit-test 500
  (do (println "Slong rational unit test: The decoded number in absolute value is > 1 when abs(numerator) > abs(denominator)")
    (prop/for-all [b0 (gen/vector gen/byte 4)
                   b1 (gen/such-that #(not= 0 (reduce + %)) (gen/vector gen/byte 4))]
      (let [res (decode-slongrational (concat b0 b1))]
        (if (> (abs (decode-slong b0)) (abs (decode-slong b1)))
          (> (abs res) 1)
          (<= (abs res) 1))))))

(defspec float32-sign-test 300
  (do (println "Float32 sign test: The decoded float has the proper sign")
    (prop/for-all [b0-pos (gen/such-that (complement neg?) gen/byte 20)
                   b0-neg (gen/such-that neg? gen/byte)
                   b1 (gen/vector gen/byte 3)]
      (and (not (neg? (decode-float32 (cons b0-pos b1))))
           (neg? (decode-float32 (cons b0-neg b1)))))))

#_(defspec float32-limit-test 500
    (do (println "Float32 limit test: The decoded float has a value between specified limits")
     (prop/for-all [b (gen/such-that #(not (zero? (reduce + %))) (gen/vector gen/byte 4))]
       (let [[b0 b1 & _] b
             exp-bin (concat (rest (encode-bits (decode-ubyte b0))) (take 1 (encode-bits (decode-ubyte b1))))
             exp (- (decode-bits exp-bin) 127)
             res (decode-float32 b)]
          (if (neg? exp)
            (and (< (float (reduce / 1 (repeat (abs exp) 2N))) (abs res))
                 (>= (reduce / 1 (repeat (inc (abs exp)) 2N)) (abs res)))
            (and (< (reduce * (repeat exp 2N)) (abs res))
                 (>= (reduce * (repeat (inc exp) 2N)) (abs res))))))))


(defspec float32-mantissa-error-test 500
  (do (println "Float32 mantissa error test: The computed manitssa of two consecutive bytes is smaller than 1.1920929E-7")
    (prop/for-all [b0 (gen/elements [0 0x80])
                   b1 (gen/such-that (complement neg?) gen/byte 20)
                   [b2 b3] (gen/vector gen/byte 2)]
      (> 1.1920929E-7 (abs (- (decode-float32 [b0 b1 b2 (inc b3)]) (decode-float32 [b0 b1 b2 b3])))))))

(defspec ascii-string-no-read-after-null 2000
  (do (println "Ascii string null byte test: The decoded string does not include substring after the null byte")
    (prop/for-all [s0 (gen/list (gen/such-that pos? gen/byte 30))
                   s1 (gen/list (gen/such-that pos? gen/byte 30))]
      (= (decode-ascii-str s0) (decode-ascii-str (concat s0 [0] s1))))))


test/org/clojars/nighcoder/binio/core_test.clj
(ns org.clojars.nighcoder.binio.core-test
  (:require clojure.test
            [clojure.test.check :as tc]
            [clojure.test.check.generators :as gen]
            [clojure.test.check.properties :as prop]
            [clojure.test.check.clojure-test :refer [defspec]]
            [org.clojars.nighcoder.binio.core :refer :all]))

(defn ubyte?
  [n]
  (and (integer? n)
       (<= 0 n 255)))

(defn abs
  [x]
  (if (neg? x)
    (- x)
    x))

(defspec bits-roundtrip-test
  (do (println "Bits roundtrip test: Any number converted to bits and back remains unchanged")
    (prop/for-all [n (gen/choose 0 255)]
      (= (decode-bits (encode-bits n)) n))))

(defspec ubyte-test
  (do (println "Ubyte test: The decoded byte is an integer between 0 and 255 (inclusive)")
    (prop/for-all [b gen/byte]
      (ubyte? (decode-ubyte b)))))

(defspec ushort-sign-test 500
  (do (println "Ushort test: The decoded number is an integer between 0 and 65.535 (inclusive)")
    (prop/for-all [b (gen/vector gen/byte 2)]
      (and (integer? (decode-ushort b))
           (<= 0 (decode-ushort b) 65535)))))

(defspec ushort-sshort-conversion-test 500
  (do (println "Ushort Sshort conversion test: The decoded sshort converted to a ushort is the same as the decoded ushort")
    (prop/for-all [[b0 b1 :as b] (gen/vector gen/byte 2)]
      (if (neg? b0)
        (= (+ 65536 (decode-sshort b)) (decode-ushort b))
        (= (decode-ushort b) (decode-sshort b))))))

(defspec ulong-test 500
  (do (println "Ulong test: The decoded number is an integer between 0 and 4.294.967.295")
    (prop/for-all [b (gen/vector gen/byte 4)]
      (and (integer? (decode-ulong b))
           (<= 0 (decode-ulong b) 4294967295)))))

(defspec ulong-sslong-conversion-test 500
  (do (println "Ulong Slong conversion test: The decoded slong converted to a ulong is the same as the decoded ulong")
    (prop/for-all [[b0 & _ :as b] (gen/vector gen/byte 4)]
      (if (neg? b0)
        (= (+ 4294967296 (decode-slong b)) (decode-ulong b))
        (= (decode-ulong b) (decode-slong b))))))

(defspec ulong-rational-pos-test 500
  (do (println "Ulong rational positive test: The decoded number is a positive rational")
    (prop/for-all [b0 (gen/vector gen/byte 4)
                   b1 (gen/such-that #(not= 0 (reduce + %)) (gen/vector gen/byte 4))]
      (let [res (decode-ulongrational (concat b0 b1))]
        (and (pos? res) (rational? res))))))

(defspec ulong-rational-unit-test 500
  (do (println "Ulong rational unit test: The decoded number is > 1 when numerator > denominator")
    (prop/for-all [b0 (gen/vector gen/byte 4)
                   b1 (gen/such-that #(not= 0 (reduce + %)) (gen/vector gen/byte 4))]
      (let [res (decode-ulongrational (concat b0 b1))]
        (if (> (decode-ulong b0) (decode-ulong b1))
          (> res 1)
          (<= res 1))))))

(defspec slong-rational-test 500
  (do (println "Slong rational test: The decoded number is rational")
    (prop/for-all [b0 (gen/vector gen/byte 4)
                   b1 (gen/such-that #(not= 0 (reduce + %)) (gen/vector gen/byte 4))]
      (let [res (decode-ulongrational (concat b0 b1))]
        (rational? res)))))

(defspec slong-rational-sign-test 500
  (do (println "Slong rational sign test: The decoded number is positive if both the numerator and denominator are either positive or negative")
    (prop/for-all [b0 (gen/vector gen/byte 4)
                   b1 (gen/such-that #(not= 0 (reduce + %)) (gen/vector gen/byte 4))]
      (let [res (decode-slongrational (concat b0 b1))]
        (if (or (and (pos? (decode-slong b0)) (pos? (decode-slong b1)))
                (and (neg? (decode-slong b0)) (neg? (decode-slong b1))))
          (pos? res)
          (neg? res))))))

(defspec slong-rational-unit-test 500
  (do (println "Slong rational unit test: The decoded number in absolute value is > 1 when abs(numerator) > abs(denominator)")
    (prop/for-all [b0 (gen/vector gen/byte 4)
                   b1 (gen/such-that #(not= 0 (reduce + %)) (gen/vector gen/byte 4))]
      (let [res (decode-slongrational (concat b0 b1))]
        (if (> (abs (decode-slong b0)) (abs (decode-slong b1)))
          (> (abs res) 1)
          (<= (abs res) 1))))))

(defspec float32-sign-test 300
  (do (println "Float32 sign test: The decoded float has the proper sign")
    (prop/for-all [b0-pos (gen/such-that (complement neg?) gen/byte 20)
                   b0-neg (gen/such-that neg? gen/byte 20)
                   b1 (gen/vector gen/byte 3)]
      (and (not (neg? (decode-float32 (cons b0-pos b1))))
           (neg? (decode-float32 (cons b0-neg b1)))))))

#_(defspec float32-limit-test 500
    (do (println "Float32 limit test: The decoded float has a value between specified limits")
     (prop/for-all [b (gen/such-that #(not (zero? (reduce + %))) (gen/vector gen/byte 4))]
       (let [[b0 b1 & _] b
             exp-bin (concat (rest (encode-bits (decode-ubyte b0))) (take 1 (encode-bits (decode-ubyte b1))))
             exp (- (decode-bits exp-bin) 127)
             res (decode-float32 b)]
          (if (neg? exp)
            (and (< (float (reduce / 1 (repeat (abs exp) 2N))) (abs res))
                 (>= (reduce / 1 (repeat (inc (abs exp)) 2N)) (abs res)))
            (and (< (reduce * (repeat exp 2N)) (abs res))
                 (>= (reduce * (repeat (inc exp) 2N)) (abs res))))))))


(defspec float32-mantissa-error-test 500
  (do (println "Float32 mantissa error test: The computed manitssa of two consecutive bytes is smaller than 1.1920929E-7")
    (prop/for-all [b0 (gen/elements [0 0x80])
                   b1 (gen/such-that (complement neg?) gen/byte 20)
                   [b2 b3] (gen/vector gen/byte 2)]
      (> 1.1920929E-7 (abs (- (decode-float32 [b0 b1 b2 (inc b3)]) (decode-float32 [b0 b1 b2 b3])))))))

(defspec ascii-string-no-read-after-null 2000
  (do (println "Ascii string null byte test: The decoded string does not include substring after the null byte")
    (prop/for-all [s0 (gen/list (gen/such-that pos? gen/byte 30))
                   s1 (gen/list (gen/such-that pos? gen/byte 30))]
      (= (decode-ascii-str s0) (decode-ascii-str (concat s0 [0] s1))))))


src/org/clojars/nighcoder/binio/core.clj
(ns org.clojars.nighcoder.binio.core
  (:require [clojure.java.io :as io])
  (:import [java.io ByteArrayOutputStream FileOutputStream IOException]))

(defn slurp-bytes
  "Attempts to open uri as input stream and copy its content to a byte array."
  [^String uri]
  (let [out (new ByteArrayOutputStream)]
    (io/copy (io/input-stream uri) out)
    (.toByteArray out)))

(defn spit-bytes
  "Saves `content` to `filename`. Passes `opt` to output stream."
  [^String filename content & opt]
  (if (bytes? content)
    (with-open [file (io/make-output-stream filename (when opt (apply hash-map opt)))]
      (.write file content))
    (throw (IOException. "Error. Must supply a byte-array to write"))))

(defn encode-bits
  "Returns a list of 8 bits (0 or 1) that represents the number `n` in base 2.
  `n` is a unsigned byte or an integer from 0 to 255 inclusive.

  Example usage:
```
  => (encode-bits 22)
  (0 0 0 1 0 1 1 0)
```"
  [n]
  (->> (repeat n)
       (map bit-and [128 64 32 16 8 4 2 1])
       (map #(if (pos? %) 1 0))))

(defn decode-bits
  "Returns the decimal value of `bin-seq`, where `bin-seq` is a list of 8 bits (0 or 1).

  Example usage:
```
  => (decode-bits [0 0 1 0 1 1 1 0])
  46
```"
  [bin-seq]
  (->> (iterate (partial * 2) 1)
       (take 8)
       reverse
       (map * bin-seq)
       (reduce +)))

(defn decode-ubyte
  "Returns the base 10 value of `b` interpreted as an unsigned byte (8 bit integer), where `b` is a **byte** or and integer from -128 to 127. "
  [b]
  (let [b (short b)]
    (if (neg? b)
      (+ b 256)
      b)))

(defn decode-ushort
  "Returns the base 10 value of a *seq of two bytes* interpreted as an unsigned short (16 bit integer).

  Example usage:
```
  => (decode-ushort [12 -30])
  3298
```"
  [[b0 b1]]
  (int (bit-or (bit-shift-left (decode-ubyte b0) 8) (decode-ubyte b1))))

(defn decode-sshort
  "Returns the base 10 value of a *seq of two bytes* interpreted as a signed short (16 bit integer).

  Example usage:
```
  => (decode-sshort [12 -30])
  3298
```"
  [[b0 b1]]
  (if (neg? b0)
    (- (inc (decode-ushort [(bit-not b0) (bit-not b1)])))
    (decode-ushort [b0 b1])))

(defn decode-ulong
  "Returns the base 10 value of a *seq of four bytes* interpreted as an unsigned long (32 bit integer).

  Example usage:
```
  => (decode-ulong [-2 -30 118 108])
  4276254316
```"
  [[b0 b1 b2 b3]]
  (long (bit-or (bit-shift-left (decode-ubyte b0) 24)
                (bit-shift-left (decode-ubyte b1) 16)
                (bit-shift-left (decode-ubyte b2) 8)
                (decode-ubyte b3))))

(defn decode-slong
  "Returns the base 10 value of a *seq of four bytes* interpreted as a signed long (32 bit integer).

  Example usage:
```
  => (decode-slong [-2 -30 118 108])
  -18712980
```"
  [[b0 b1 b2 b3]]
  (if (neg? b0)
    (- (inc (decode-ulong (map bit-not [b0 b1 b2 b3]))))
    (decode-ulong [b0 b1 b2 b3])))

(defn decode-ulongrational
  "Returns the rational value of a *seq of eight bytes* interpreted as two unsigned longs (32 bit integers).
  First four bytes form the numerator and the last four bytes form the denominator.

  Example usage:
```
  => (decode-ulongrational [34 41 -113 80 -33 -21 9 33])
  573149008/3756722465
```"
  [[b0 b1 b2 b3 b4 b5 b6 b7]]
  (/ (decode-ulong [b0 b1 b2 b3])
     (decode-ulong [b4 b5 b6 b7])))

(defn decode-slongrational
  "Returns the rational value of a *seq of eight bytes* interpreted as two signed longs (32 bit integers).
  First four bytes form the numerator and the last four bytes form the denominator.

  Example usage:
```
  => (decode-slongrational [34 41 -113 80 -33 -21 9 33])
  -573149008/538244831
```"
  [[b0 b1 b2 b3 b4 b5 b6 b7]]
  (/ (decode-slong [b0 b1 b2 b3])
     (decode-slong [b4 b5 b6 b7])))

(defn hex [n]
  (format "0x%02X" n))

(defn decode-float32
  "Returns the decimal value of a *seq of four bytes* interpred as single precision floating number."
  [[b0 b1 b2 b3]]
  (let [m_exp (take 24 (iterate #(float (/ % 2)) 1))
        b0 (encode-bits (decode-ubyte b0))
        b1 (encode-bits (decode-ubyte b1))
        b2 (encode-bits (decode-ubyte b2))
        b3 (encode-bits (decode-ubyte b3))
        bits (concat b0 b1 b2 b3)
        sign (first bits)
        exponent (- (decode-bits (take 8 (rest bits))) 127)
        mantissa (cons 1 (drop 9 bits))]
     (* (reduce * (repeat sign -1))
        (reduce + (map * mantissa m_exp))
        (if (neg? exponent)
          (reduce / 1 (repeat (- exponent) 2N))
          (reduce * (repeat exponent 2N))))))

(defn decode-ascii-str
  "Returns the contents of `b-seq` encoded as ascii string up to first occurence of 0 or the end of sequence."
  [b-seq]
  (->> b-seq (take-while (complement zero?)) (map char) (reduce str)))
